src/index.js

```diff
@@ -114,46 +137,50 @@ async function fetchWithNetworkEvents(el: Element, url: string, options: Request
 }
 
 async function check(autoCheckElement: AutoCheckElement) {
-  const src = autoCheckElement.src
-  if (!src) {
-    throw new Error('missing src')
-  }
-  const csrf = autoCheckElement.csrf
-  if (!csrf) {
-    throw new Error('missing csrf')
-  }
   const input = autoCheckElement.input
-  if (!input) return
+  if (!input) {
+    return
+  }
 
+  const src = autoCheckElement.src
+  const csrf = autoCheckElement.csrf
   const state = states.get(autoCheckElement)
-  if (!state) return
+
+  // If some attributes are missing we want to exit early and make sure that the element is valid.
+  if (!src || !csrf || !state) {
+    if (autoCheckElement.required) {
+      input.setCustomValidity('')
+    }
+    input.dispatchEvent(new CustomEvent('auto-check-complete', {bubbles: true}))
+    return
+  }
 
   const body = new FormData()
   body.append('authenticity_token', csrf)
   body.append('value', input.value)
 
   const id = body.entries ? [...body.entries()].sort().toString() : null
-  if (id && id === state.previousValue) return
+  if (id && id === state.previousValue) {
+    input.dispatchEvent(new CustomEvent('auto-check-complete', {bubbles: true}))
```

Comment:
Hmm consider a use case like this:

```js
on('auto-check-start', 'auto-check', event => {
  removeCurrentState() // since we are re-checking now
  event.target.classList.remove('is-loading')
})

on('auto-check-complete', 'auto-check', event => {
  event.target.classList.remove('is-loading')
})
```

By the end of this no-op the previous success/error state is lost. 

`removeCurrentState` would need to be called just so the input reflects that a new check is being kicked off. Unless we expected users to only set loading on `auto-check-send`? But this would create an invalid input with potentially a success check mark.

Reply:
Could you elaborate more? I'm not 100% following with the given use case.

Reply:
In our own code: https://github.com/github/github/blob/master/app/assets/modules/github/behaviors/autocheck.js

Current behavior— 

Say, before a new `change` event happens, we have already an `<auto-check class=“is-error” required>`.

The code:

- `loadstart`: remove state classes, set loading classes: `<auto-check class=“is-loading” required>`
- `loadend`: remove loading class: `<auto-check required>`

Now, the value did not change at this `change` event, therefore `loadstart` was never fired, so element stays as `<auto-check class=“is-error” required>`

But with this new change—

The “start” of auto-check isn’t “loadstart” anymore, but “auto-check-start”, the UI should reflect this as well. So this happens:

Starting from `<auto-check class=“is-error” required>` again...

- auto-check-start: remove state classes, set loading classes: `<auto-check class=“is-loading” required>`
- auto-check-complete: remove loading class: `<auto-check required>`

Now, even though the value did not change, both of these events were fired, and we ended up with a valid but incorrect `<auto-check required>`.

—

And say, if we don’t change how state classes are added, keeping them with `loadstart` and `loadend`...

Before a new `change` event we have `<auto-check class=“is-success” required>`, `auto-check-start` would set input as invalid on change, but because `loadstart` hasn’t happened, input still has a green checkmark showing it is valid, when in reality it isn’t.

Reply:
I added a clarification to the examples in e1d38210. Network progress state in the user interface, like spinners, must be managed by the network lifecycle events.

Proposed behavior of auto-check-* events:

- `auto-check-send` must be dispatched to allow for customizing the request. If any prerequisite to the request is not met, auto-check-send must not be dispatched. This includes missing attributes or an empty input value.
- `auto-check-complete` must be dispatched only if `auto-check-send` is dispatched to complete the lifecycle
- `auto-check-complete` must be dispatched after `auto-check-success` or `auto-check-error` to complete the lifecycle
- `auto-check-success` must be dispatched if the server response passed validation
- `auto-check-error` must be dispatched if the server response failed validation. This does not include network failures. If no server response is received, the lifecycle must be `auto-check-send` -> `auto-check-complete`.

Applying those rules, the above change that dispatches `auto-check-complete` when the form is identical to the previous form should be reverted. An `auto-check-complete` event has already been dispatched for the prior form and no `auto-check-send` event will be dispatched for this duplicate form. It has failed a prerequisite to being sent to the server.